### PR TITLE
[VirtualOutputBackend] Proper error reporting for FileSystem errors

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Return proper llvm::Error from virtual output backend when it discovers file system error when keep/discard. Instead of triggering the fatal error in raw_fd_ostream, report the error_code as llvm::Error to caller.

rdar://107855318

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
Don't know how to write a test case that actually trigger a file system error.

@swift-ci please test

File: llvm/lib/Support/VirtualOutputBackends.cpp [hunk 1]
```diff
@@ -381,8 +395,8 @@ Error OnDiskOutputFile::keep() {
 
 Error OnDiskOutputFile::discard() {
   // Destroy the streams to flush them.
-  BufferOS.reset();
-  FileOS.reset();
+  if (auto E = reset())
+    return E;
```
Comment (benlangmuir) [inline llvm/lib/Support/VirtualOutputBackends.cpp:399]: Might make sense to wrap in `llvm::createFileError` so we can add the path.

**Review by benlangmuir — APPROVED** [review]
For test: maybe you could create a directory that does not have write permission then try to have an output inside it? But I don't feel like you need to block on it.

**Comment by cachemeifyoucan** [comment]
Not really. That is an error on creation, not closing. The only one I found is do an illegal seek out of bound but that will trigger an assertion. Maybe also trigger with write a buffer bigger than disk space but that is not a good test.

**Pull request merged**